### ISSUE TITLE
Cat form does not increase agility (only AP)

### ISSUE THREAD

**Issue by ghost** [mangosArchives/databaseZero_Rel19#25]
Normal
![WoWScrnShot_020813_213501](https://f.cloud.github.com/assets/3449551/140817/46546fe8-722f-11e2-8a75-1ccc7acd0f8e.jpg)

Cat Form
![WoWScrnShot_020813_213509](https://f.cloud.github.com/assets/3449551/140819/5471d354-722f-11e2-8e1a-8c41203b2e19.jpg)

**Comment by Salja** [comment]
Its Core Issues i have maked a fix pls testing anf give feedback the fix is on the server / testing branch
https://github.com/mangoszero/server/commit/c3bc4fdd0382280f033cd9ef0bffb48b725f096b

**Comment by The-Great-Sephiroth** [comment]
Agility is not supposed to be increased in cat form. The text is "Increases attack power by 40 plus agility". In other words if you have 100 agility and 300 AP normally, switching to cat form would make your AP 440. Why? 40 bonus AP plus 100 agility equals 140 bonus AP. This is not a bug.

View the link below to see this. Again, this is not a bug. Agility does not get increased going to cat form in 1.12!
http://db.valkyrie-wow.com/?spell=768

**Comment by Salja** [comment]
so is working yet all right and we can close here ?

**Comment by The-Great-Sephiroth** [comment]
According to the two pictures, no. He has 139 agility and 256 AP in normal form. In cat form the AP goes to 635. The problem I see is too much AP being applied. 139 agility + 40 bonus AP + 256 base AP equals 435, not 635. So the reality is that too much AP is being applied somehow. This is really a simple formula.

int bonus = player.agility + 40;
player.ap = player.ap + bonus;

Not sure if those members are correct but you get the idea. Also, this should be tested with no armor or items on. He may have a 200 AP bonus when in cat form going on through items. Either way the calculation is way off.

**Comment by The-Great-Sephiroth** [comment]
I now have a lv21 Tauren druid. I stripped her naked and removed her weapon. Her agility is 25 and base attack power is 54. Based on adding 40 attack power plus agility, that means my AP should go up by 65 (40 plus my 25 agility). That would put me at 119 AP. Instead I somehow went to 121 AP. Granted, it is only 2 AP extra, but this is also only lv21. Does that wind up being 200 extra AP at lv60? No clue, yet. Either way, the formula is not right. I am assuming that this is a core bug and not a DB issue as well.

**Comment by Salja** [comment]
can you test and make this commit revert ? https://github.com/mangoszero/server/commit/c3bc4fdd0382280f033cd9ef0bffb48b725f096b

**Comment by The-Great-Sephiroth** [comment]
I can. I will recompile today and let you know. Is it possible to use git to apply this patch to the stable branch or do I have to use the testing branch?

**Comment by Salja** [comment]
you need to revert  git revert c3bc4fdd0382280f033cd9ef0bffb48b725f096b

**Comment by The-Great-Sephiroth** [comment]
Alright, give me some time. I will let you know.

**Comment by Salja** [comment]
ok thanks you the formel man XD

i will look today of your quest bug reports btw its easy you make a account on test server forum and use the quest tracker for reports than i found some faster http://project-zero.eu/index.php?page=tracker

**Comment by The-Great-Sephiroth** [comment]
Problem. It keeps telling me that c3bc4fd is an unknown revision, both in testing and master branches. I can manually make the change by hand and compile if you'd like.

**Comment by Salja** [comment]
yes its easyer change by hand back

**Comment by The-Great-Sephiroth** [comment]
Alright, building it now, I will let you know if this fixes the issue.

**Comment by The-Great-Sephiroth** [comment]
Salja, looking at the sources this change will do nothing. The piece that was changed has to do with "Predatory Strikes". It will not affect the output either way. The code should be changed to the following unless I am mistaken.

``` C
case FORM_CAT:
  //val2 = getLevel() * mLevelMult + GetStat(STAT_STRENGTH) * 2.0f + GetStat(STAT_AGILITY) - 20.0f; break;
  val2 = (getLevel() * mLevelMult) + (GetStat(STAT_STRENGTH) * 2.0f) + GetStat(STAT_AGILITY) + 20.0f; break;
```

Again, the tooltip says it increases AP forty plus agility. The formula above does exactly that. Also, why are we subtracting 20 AP at the end of the formula by default?

_UPDATE_

I see where predatory strikes comes into play. The correct code is above, unless that subtraction of twenty is required.

_UPDATE_

Alright I am not following something. The code above gave me the correct 119 AP, but somehow the game is giving me +42AP on top of that. Is this code explained somewhere? I see at the bottom of this function that it sets val2 as the BASE value, which is being set properly to 119 in cat form for my character. Why am I getting an additional 42AP though?

**Comment by The-Great-Sephiroth** [comment]
Alright I am trying to dig in and fix this but I need a few things explained.

``` C
SetModifierValue(unitMod, BASE_VALUE, val2); //This line apparently sets my base AP based on the formula above

float base_attPower  = GetModifierValue(unitMod, BASE_VALUE) * GetModifierValue(unitMod, BASE_PCT); //No clue what this does
float attPowerMod = GetModifierValue(unitMod, TOTAL_VALUE); //How is this mod calculated and why?
float attPowerMultiplier = GetModifierValue(unitMod, TOTAL_PCT) - 1.0f; //What is this multiplier used for?

//No idea what any of these do
SetInt32Value(index, (uint32)base_attPower); 
SetInt32Value(index_mod, (uint32)attPowerMod);
SetFloatValue(index_mult, attPowerMultiplier);
```

Can you explain these lines of code to me so that I may understand what is going on? I can fix this easily if I understand the code.

_UPDATE_

Alright, found some old screenshots and I see another issue. The base AP should always be calculated the same way. The bonus from changing into cat form should be green AP. In other words if I have a normal base AP of 100 and I have 60 agility, then when I change into cat form it should be "AP: 100 + 100", not 200 base AP. I get the 100 bonus AP from the stock 40 AP bonus and my 60 agility. This is probably where a good chunk of our problem is.

**Comment by The-Great-Sephiroth** [comment]
Also, I just found somebody else explaining the bonus on thottbot from way back in the day. Link follows. Agility does NOT get increased in cat form.

http://web.archive.org/web/20070514135655/http://www.thottbot.com/s768

**Comment by danielsreichenbach** [comment]
This requires a bit of research, and thus will be scheduled for release 2.0.8, since it can not be resolved for 2.0.7.

**Comment by danielsreichenbach** [comment]
This is a server topic, thus issue closed.
